usr/src/cmd/zdb/zdb.c

```diff
@@ -2142,6 +2175,117 @@ dump_label_uberblocks(vdev_label_t *lbl, uint64_t ashift)
 }
 
 static int
+dump_specific_path(objset_t *os, uint64_t obj)
+{
+	int print_header = 1;
+
+	dump_object(os, obj, dump_opt['v'], &print_header);
+
+	return (0);
+}
+
+static int
+dump_specific_dir(objset_t *os, uint64_t obj, char *topname, char *name)
+{
+	int err;
+	uint64_t child_obj;
+	char *slash = strchr(name, '/');
+	char *namep = name;
+	dmu_buf_t *db;
+	dmu_object_info_t doi;
+
+	if (slash != NULL)
+		*slash = '\0';
+	err = zap_lookup(os, obj, namep, 8, 1, &child_obj);
+	if (slash != NULL)
+		*slash = '/';
+
+	if (err != 0) {
+		(void) fprintf(stderr, "failed to lookup '%s': %s\n",
+		    name, strerror(err));
+		return (err);
+	}
+
+	child_obj = ZFS_DIRENT_OBJ(child_obj);
+	err = sa_buf_hold(os, child_obj, FTAG, &db);
+	if (err != 0) {
+		(void) fprintf(stderr,
+		    "failed to get SA dbuf for obj %llu: %s\n",
+		    (u_longlong_t)obj, strerror(err));
+		return (EINVAL);
+	}
+	dmu_object_info_from_db(db, &doi);
+	sa_buf_rele(db, FTAG);
+
+	if (dump_opt['v'] > 6) {
+		(void) fprintf(stderr, "%s: obj %llu name '%s' type %d "
+		    "bonustype %d\n", __func__, (u_longlong_t)obj, name,
+		    doi.doi_type, doi.doi_bonus_type);
+	}
+
+	if (doi.doi_bonus_type != DMU_OT_SA &&
+	    doi.doi_bonus_type != DMU_OT_ZNODE) {
+		(void) fprintf(stderr, "invalid bonus type %d for obj %llu\n",
+		    doi.doi_bonus_type, (u_longlong_t)obj);
+		return (EINVAL);
+	}
+
+	switch (doi.doi_type) {
+	case DMU_OT_PLAIN_FILE_CONTENTS:
+		if (slash != NULL) {
+			*slash = '\0';
+			(void) fprintf(stderr, "error: path '/%s' specifies "
+			    "a file\n", topname);
+			break;
+		}
+		return (dump_specific_path(os, child_obj));
+	case DMU_OT_DIRECTORY_CONTENTS:
+		if (slash == NULL || *(slash + 1) == '\0') {
+			if (dump_opt['v'] > 6)
+				(void) fprintf(stderr, "obj %llu name '%s' is "
+				    "a directory at the end of the chain\n",
+				    (u_longlong_t)obj, name);
+			/* Directory is the end of the chain */
+			return (dump_specific_path(os, child_obj));
+		}
+		return (dump_specific_dir(os, child_obj, topname, slash + 1));
+	default:
+		(void) fprintf(stderr, "object %llu has non-file/directory "
+		    "type %d\n", (u_longlong_t)obj, doi.doi_type);
+		break;
+	}
+
+	return (EINVAL);
+}
+
+/*
+ * Dump the blocks for the specified object.
+ * Args: <dataset> <path>
+ */
+static int
+dump_specific_object(char *ds, char *path)
```

Comment:
Modified the function names, now we have dump_path(), calling dump_path_impl(), which is recursive.